File: ch06-unix/01_shell_demo/playground.mjs
import { spawn, exec } from 'node:child_process'

console.log('命令行所有参数', process.argv)
console.log('进程id', process.pid)
console.log('父进程id', process.ppid)

// spawn() 会打开一个子进程
const subprocess = spawn(
    'echo',
    ['hello world'],
    {
        // env: process.env, // 默认将环境变量传递给子进程
        // 一般变量用小写 环境变量一般用大写
        env: { MODE: 'development' }, // 也可以将指定环境变量传递给子进程
    }
)

subprocess.stdout.on('data', (data) => {
    console.log(data.toString('utf-8'))
})

exec(
    "echo 'something string' | tr ' ' '\n'",
    {
        shell: '/bin/zsh'
    },
    (error, stdout, stderr) => {
        if (error) {
            console.error(error)
            return
        }
    }
)

File: ch06-unix/01_shell_demo/playground.c
#include <stdio.h>
#include <unistd.h>
#include <stdlib.h>

// node process.argv 实际类似下面c程序
// process.pid -> getpid()
// process.ppid -> getppid()
int main(int argc, char *argv[]) {
    for (int i = 0; i < argc; i++) {
        printf("argv number %d is: %s\n", i, argv[i]);
    }

    printf("Process ID: %d\n", getpid());
    printf("Parent Process ID: %d\n", getppid());

    // 获取环境变量
    printf("MODE env is: %s", getenv("MODE"));
    return 0;
}